### DISCUSSION TITLE
XLS-0044d Physical and Logical Redesign

### DISCUSSION THREAD

**Discussion by thejohnfreeman** [XRPLF/XRPL-Standards#121] [XLS Proposals]
I've spent some time working on https://github.com/XRPLF/rippled/issues/4606. It has given me some more insights, and I am returning with revisions to the plan.

I would like to see more changes than just a physical redesign, to include renaming logical targets and binaries, and moving more software components from the executable `rippled` into the library `libxrpl`. Let me start with the benefits I am trying to achieve:

- **Easier prototyping of new tools.** When I first started working on forward ledger replay, I wanted to quickly write a small tool to connect to a local database, load a ledger, and compute statistics on it, including things like minimum and maximum depths of the tree and minimum and maximum object sizes. But to do so, I would have to edit `rippled` because some of the components I needed (`NodeStore`, `Ledger`, `SHAMap`) are not in `libxrpl`. I'd like to get to the point where a light client for the XRPL network protocol, like [xrpl-peermon](https://github.com/RichardAH/xrpl-peermon), can be written in just a few lines on top of `libxrpl`.
- **Better tests and benchmarks.** There are many large pieces of `rippled` that are impossible to test in isolation because of high coupling. They need to be isolated from each other and separately testable. Right now, the only way to reproducibly test some large pieces like forward ledger replay is to run all of `rippled`, with a carefully constructed configuration file, and in a special environment with a network of test machines. Many developers do not have access to resources like this. I want to make it possible to execute benchmarks locallly.
- **Levelization on sight.** We can make it possible for developers to easily spot improper dependencies based just on the include paths. Nothing in `libxrpl` should depend on anything in `rippled`, but when they live in the same directory, it is hard to notice when that happens.
- **Faster iterations during development.** When it comes to rebuilding, we can split translation units into more libraries. When it comes to retesting, we can move tests out of `rippled` into separate binaries. Both of these changes create binaries from smaller sets of object files, reducing the time and memory it takes to link them.
- **Ease contributions.** Newcomers and old-timers alike benefit when a codebase follows popular conventions. There are no questions about where new files should go. We don't have to worry about developers forgetting to add new files to source lists in the CMake. Everything is installed correctly, the build rules are simple, and the `include` directory in the codebase (before installation) looks exactly how it will look after installation.

The physical and logical redesign described below achieves two of these goals (levelization on sight, ease contributions) immediately. The rest require decoupling components from each other and moving them into `libxrpl`, and this redesign is a first step to enable and ease that transition. I believe these changes can be executed as multiple, _independent_ phases (meaning no one has to wait on any other):

1. Rearrange first-party sources: `include/xrpl`, `src/libxrpl`, `src/xrpld`.

    These changes will be disruptive to developers (people who write code for `rippled`), but not to consumers (projects that depend on `libxrpl`) or operators (people who run `rippled`). My plan is to have a script that makes these changes for you on whatever branches you have in-progress. The process would be this:

    - In the upstream, there will be a commit `A` before these changes, and the first commit `B`, child of `A`, that includes these changes.
    - You merge your branch with `A` and resolve conflicts yourself. This is the catch-up step. Conflicts are the ones you would have had to resolve anyway even if `B` did not exist.
    - You run the script. There should be no errors. It will have been developed with cooperation from multiple contributors and tested in multiple scenarios. **I will personally handle any issues people experience with this step.** The script will move files, delete some directories left empty, change `#includes`, and overwrite build files.
    - You commit the above changes. You can `git diff` with the upstream and see what should be only your changes in the branch.
    - You `git merge` with the upstream. This should not make any changes, but create a commit with the upstream as a parent.
    - You can continue your work as before.

2. Move protobufs into `libxrpl`. This should be disruptive to no one, but allow consumers like [Clio](https://github.com/XRPLF/clio) to reuse the protobufs.
3. Rearrange third-party sources: `external/secp256k1`, `external/ed25519`
4. Rearrange CMake files: `cmake/`
5. Move container and packaging files to a separate repository.
6. Remove files that appear to be unused: `json/TODO.md`, `json/impl/LICENSE` (we may just move this one into the top-level `LICENSE` file), `json/impl/version`, ...?
7. Rename the `rippled` executable and CMake target to `xrpld` (or perhaps something else). This will be the most disruptive to operators.
8. Rename the `ripple` namespace to `xrpl`. This will be the most disruptive to `libxrpl` consumers.

I am seeking buy-in from the community for all of these changes. I anticipate 7 will be the most contentious. Please share your thoughts here.

**Comment by wojake** [comment]
Great proposal and I think we're all for it, and I agree that `#7` is volatile, so I suppose we'd move in milestones and emit enough warnings to inform sufficient operators & users.

How do you plan on executing this proposal?

**Reply by thejohnfreeman** [reply]
Each of these will be separate PRs, in no particular order, as I find the time and motivation. Implementations are pretty straightforward:

1. A process I laid out in the draft.
2. Small change to the CMake. `git mv` a few files.
3. `git mv` two directories.
4. `git mv` one directory.
5. `git rm` one directory (and open a new repository with the files we want to keep).
6. `git rm` a few files.
7. Small change to the CMake. Change the documentation.
8. `s/namespace ripple/namespace xrpl` and `s/ripple::/xrpl::/`.

**Comment by ObiajuluM** [comment]
Out of curiosity, this proposal will also apply to SDKs right

**Reply by ckniffen** [reply]
I don't think this will impact client libraries unless they use libxrpl.  I don't know of any that do.

**Comment by intelliot** [comment]
Let's punt (pause) on `#7` and move forward on the rest of the changes. In other words, the overall project name will remain `rippled` for now.

**Reply by mvadari** [reply]
Is it possible to have an alias called `xrpld` or something? So that you can have the best of both worlds (`rippled` to have backwards compatibility, `xrpld` for better naming)

**Reply by thejohnfreeman** [reply]
It is possible, and that's much like what I've done for `libxrpl` already, but I did that because there are few consumers referencing the other name, and they are likely all within Ripple. Before I made the library easier to use for new consumers, I took the opportunity to introduce the name I wanted them to use. Plus, I plan to eventually deprecate and retire the old name. The cat is already out of the bag for `rippled`, though. I hesitate to introduce an alternative name unless it becomes clear that it will become the official name and replace the name `rippled`.

**Reply by wojake** [reply]
I do think we should focus on other things before we consider renaming as well.

**Comment by RichardAH** [comment]
Modifying the source tree structure will make rippled PR-incompatible with forks such as Xahau. (No, we're not going to run this script.)

Making breaking changes to the structure of a heavily forked repo is usually inadvisable. Part of the point of open source software is PR portability.

**Reply by thejohnfreeman** [reply]
PRs can still be ported by using the script. Why won't you run it?

**Reply by RichardAH** [reply]
We have branches, private repos, private amendments, so do other people. We're not going to introduce a frictional breaking change into the middle of this for almost no benefit.

Incidentally if I write a script that makes breaking structural changes and insist you run it, would you have any issues with that? Why wouldn't you run that?

BTW the build system for https://github.com/ripple/validator-keys-tool is still broken since conan and no one has fixed it.

**Reply by thejohnfreeman** [reply]
Yes, of course I'd run it if you hosted a discussion here documenting the benefits and getting buy-in. Would you run your own script?

validator-keys-tool isn't on my radar, but now that you've brought it to my attention, I can fix it in a jiffy. That ok with you?

**Reply by thejohnfreeman** [reply]
I stand corrected. I fixed the build to use Conan in April of this year, on someone's request, and pushed the changes to the `conan` [branch](https://github.com/thejohnfreeman/validator-keys-tool/tree/conan) on my fork. I'll tidy them up now and open a PR.

**Comment by mvadari** [comment]
@thejohnfreeman should this spec be moved into PR and kept in the repo or should it be closed? Note that this does not mean that the spec has been accepted/implemented/finalized.

I will move it to PR by copy-pasting the existing draft if there is no response in the next week.

**Reply by thejohnfreeman** [reply]
I think it can be closed. Everything I wanted to accomplish in this proposal has landed in `develop` by now. Thank you!
